Signal/src/ViewControllers/ProfileViewController.m

```diff
@@ -204,14 +244,44 @@ - (void)setHasUnsavedChanges:(BOOL)hasUnsavedChanges
 {
     _hasUnsavedChanges = hasUnsavedChanges;
 
-    if (hasUnsavedChanges) {
-        self.navigationItem.rightBarButtonItem = (self.hasUnsavedChanges
-                ? [[UIBarButtonItem alloc] initWithTitle:NSLocalizedString(@"EDIT_GROUP_UPDATE_BUTTON",
-                                                             @"The title for the 'update group' button.")
-                                                   style:UIBarButtonItemStylePlain
-                                                  target:self
-                                                  action:@selector(updatePressed)]
-                : nil);
+    [self updateNavigationItem];
+}
+
+- (void)updateNavigationItem
+{
+    // The navigation bar is hidden in the registration workflow.
+    if (self.navigationController.navigationBarHidden) {
+        [self.navigationController setNavigationBarHidden:NO animated:YES];
+    }
+
+    // Always display a left item to leave the view without making changes.
+    // This might be a "back", "skip" or "cancel" button depending on the
+    // context.
+    switch (self.profileViewMode) {
+        case ProfileViewMode_AppSettings:
+            self.navigationItem.leftBarButtonItem =
+                [self createOWSBackButtonWithTarget:self selector:@selector(backOrSkipButtonPressed)];
+            break;
+        case ProfileViewMode_UpgradeOrNag:
+            self.navigationItem.leftBarButtonItem =
+                [[UIBarButtonItem alloc] initWithBarButtonSystemItem:UIBarButtonSystemItemCancel
+                                                              target:self
+                                                              action:@selector(backOrSkipButtonPressed)];
+            break;
+        case ProfileViewMode_Registration:
+            self.navigationItem.leftBarButtonItem = [[UIBarButtonItem alloc]
+                initWithTitle:NSLocalizedString(@"NAVIGATION_ITEM_SKIP_BUTTON", @"A button to skip a view.")
+                        style:UIBarButtonItemStylePlain
+                       target:self
+                       action:@selector(backOrSkipButtonPressed)];
+            break;
+    }
+    if (self.hasUnsavedChanges) {
+        // If we have a unsaved changes, right item should be a "save" button.
```

Comment:
Our registration flow already has the problem of wanting to use big, bottom-aligned buttons for "next/done", which is already tricky since these views have text input, and when the keyboard is up there's very little screen real estate left.  

I didn't see how to solve this problem in the profile view - we want to show the avatar large, for example.  And we haven't even added yet:

* A brief description of when your profile is visible and to who, with a link to more info.
* Any text about whether or your profile name should be your real name, whether or not it needs to be unique, etc.

![screen shot 2017-08-17 at 11 39 36 am](https://user-images.githubusercontent.com/625803/29420748-e726d420-8340-11e7-948a-6fa4973125c0.png)
![screen shot 2017-08-17 at 11 38 46 am](https://user-images.githubusercontent.com/625803/29420749-e72c3a96-8340-11e7-9436-473f9e733927.png)